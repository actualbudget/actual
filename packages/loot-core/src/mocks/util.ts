import { join, dirname, basename } from 'path';

import snapshotDiff from 'snapshot-diff';

export function expectSnapshotWithDiffer(
  initialValue,
  { onlyUpdates }: { onlyUpdates? } = {},
) {
  let currentValue = initialValue;
  if (!onlyUpdates) {
    expect(initialValue).toMatchSnapshot();
  }
  return {
    expectToMatchDiff: value => {
      expect(snapshotDiff(currentValue, value)).toMatchSnapshot();
      currentValue = value;
    },
  };
}

export function getFixtures(filename) {
  join(
    dirname(filename),
    '__fixtures__',
    basename(filename).replace(/\.[^.]+.js/, '.fixtures.js'),
  );
}

export function debugDOM(node) {
  function debugDOM(node, indent = 0) {
    let str = '';
    if (node) {
      str += ' '.repeat(indent);
      if (node.tagName) {
        str +=
          node.tagName.toLowerCase() +
          ' ' +
          (node.tabIndex || '') +
          (node.dataset['testid'] ? ' ' + node.dataset['testid'] : '') +
          '\n';
      } else {
        str += node.textContent + '\n';
      }

      for (let child of node.childNodes) {
        str += debugDOM(child, indent + 2);
      }
    }
    return str;
  }

  return debugDOM(node);
}
