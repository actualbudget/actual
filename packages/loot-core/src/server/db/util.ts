export async function incrFetch(
  runQuery,
  terms,
  compare,
  makeQuery,
  params = [],
) {
  let pageCount = 500;
  let results = [];

  let fetchedIds = new Set();

  for (let i = 0; i < terms.length; i += pageCount) {
    let slice = terms.slice(i, i + pageCount).filter(id => !fetchedIds.has(id));
    if (slice.length > 0) {
      let filter = slice.map(id => compare(id)).join(' OR ');
      let query = makeQuery('(' + filter + ')');

      let rows = await runQuery(query, params, true);
      fetchedIds = new Set([...fetchedIds, ...slice]);
      results = results.concat(rows);
    }
  }

  return results;
}

export function whereIn(ids: string[], field: string) {
  let ids2 = [...new Set(ids)];
  // eslint-disable-next-line rulesdir/typography
  let filter = `${field} IN (` + ids2.map(id => `'${id}'`).join(',') + ')';
  return filter;
}
