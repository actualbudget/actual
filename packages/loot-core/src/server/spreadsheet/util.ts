export function unresolveName(name) {
  let idx = name.indexOf('!');
  if (idx !== -1) {
    return {
      sheet: name.slice(0, idx),
      name: name.slice(idx + 1),
    };
  }
  return { sheet: null, name };
}

export function resolveName(sheet: string, name: string): string {
  return sheet + '!' + name;
}
