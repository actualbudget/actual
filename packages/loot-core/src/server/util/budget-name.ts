import { v4 as uuidv4 } from 'uuid';

import * as fs from '../../platform/server/fs';

export async function uniqueFileName(existingFiles) {
  let initialName = 'My Finances';
  let idx = 1;

  // If there is a conflict, keep appending an index until there is no
  // conflict and we have a unique name
  let newName = initialName;
  while (existingFiles.find(file => file.name === newName)) {
    newName = `${initialName} ${idx}`;
    idx++;
  }

  return newName;
}

export async function idFromFileName(name) {
  let id = name.replace(/( |[^A-Za-z0-9])/g, '-') + '-' + uuidv4().slice(0, 7);

  // Make sure the id is unique. There's a chance one could already
  // exist (although very unlikely now that we append unique
  // characters onto the id)
  let index = 0;

  let budgetDir = fs.getBudgetDir(id);
  while (await fs.exists(budgetDir)) {
    index++;
    budgetDir = fs.getBudgetDir(id + index.toString());
  }

  // If a suffix was added, update the id
  if (index > 0) {
    id = id + index.toString();
  }

  return id;
}
