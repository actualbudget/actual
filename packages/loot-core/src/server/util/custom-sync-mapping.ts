export type Mappings = Map<string, Map<string, string>>;

export const mappingsToString = (mapping: Mappings): string =>
  JSON.stringify(
    Object.fromEntries(
      [...mapping.entries()].map(([key, value]) => [
        key,
        Object.fromEntries(value),
      ]),
    ),
  );

export const mappingsFromString = (str: string): Mappings => {
  try {
    const parsed = JSON.parse(str);
    if (typeof parsed !== 'object' || parsed === null) {
      throw new Error('Invalid mapping format');
    }
    return new Map(
      Object.entries(parsed).map(([key, value]) => [
        key,
        new Map(Object.entries(value as object)),
      ]),
    );
  } catch (e) {
    const message = e instanceof Error ? e.message : e;
    throw new Error(`Failed to parse mapping: ${message}`);
  }
};

export const defaultMappings: Mappings = new Map([
  [
    'payment',
    new Map([
      ['date', 'date'],
      ['payee', 'payeeName'],
      ['notes', 'notes'],
    ]),
  ],
  [
    'deposit',
    new Map([
      ['date', 'date'],
      ['payee', 'payeeName'],
      ['notes', 'notes'],
    ]),
  ],
]);
