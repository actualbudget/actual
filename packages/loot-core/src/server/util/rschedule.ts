import '@rschedule/standard-date-adapter/setup';
import '@rschedule/json-tools/Schedule';

export * from '@rschedule/standard-date-adapter';
export * from '@rschedule/core';
export * from '@rschedule/core/generators';
