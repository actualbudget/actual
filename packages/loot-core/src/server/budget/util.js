import { number } from '../spreadsheet/globals';

export { number } from '../spreadsheet/globals';

export function sumAmounts(...amounts) {
  return amounts.reduce((total, amount) => {
    return total + number(amount);
  }, 0);
}

export function flatten2(arr) {
  return Array.prototype.concat.apply([], arr);
}

export function unflatten2(arr) {
  let res = [];
  for (let i = 0; i < arr.length; i += 2) {
    res.push([arr[i], arr[i + 1]]);
  }
  return res;
}

// Note that we don't restrict values to `Number.MIN_SAFE_INTEGER <= value <= Number.MAX_SAFE_INTEGER`
// where `Number.MAX_SAFE_INTEGER == 2^53 - 1` but a smaller range over `-(2^43-1) <= value <= 2^43 - 1`.
// This ensure that the number is accurate not just for the integer component but for 3 decimal places also.
//
// This gives us the guarantee that can use `safeNumber` on number whether they are unscaled user inputs
// or they have been converted to integers (using `amountToInteger`).

const MAX_SAFE_NUMBER = 2 ** 43 - 1;
const MIN_SAFE_NUMBER = -MAX_SAFE_NUMBER;

export function safeNumber(value) {
  value = number(value);
  if (value > MAX_SAFE_NUMBER || value < MIN_SAFE_NUMBER) {
    throw new Error(
      "Can't safely perform arithmetic with number: " + JSON.stringify(value)
    );
  }
  return value;
}
