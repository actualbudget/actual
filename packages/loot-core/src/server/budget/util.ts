import { safeNumber } from '../../shared/util';
import { number } from '../spreadsheet/globals';

export { number } from '../spreadsheet/globals';

export function sumAmounts(...amounts) {
  return safeNumber(
    amounts.reduce((total, amount) => {
      return total + number(amount);
    }, 0),
  );
}

export function flatten2(arr) {
  return Array.prototype.concat.apply([], arr);
}

export function unflatten2(arr) {
  let res = [];
  for (let i = 0; i < arr.length; i += 2) {
    res.push([arr[i], arr[i + 1]]);
  }
  return res;
}
