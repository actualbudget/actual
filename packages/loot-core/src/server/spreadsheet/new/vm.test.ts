import { unresolveName } from '../util';

import VM from './vm';

const db = {
  runQuery: sql => {
    return Promise.resolve([{ 'sum(t.amount)': 1000 }]);
  },
};

function makeScopes(vars) {
  return {
    getVariable: resolvedName => {
      const { name } = unresolveName(resolvedName);

      if (vars[resolvedName] !== undefined) {
        return vars[resolvedName];
      } else if (vars[name] !== undefined) {
        return vars[name];
      }

      throw new Error(`“${resolvedName}” is not defined`);
    },

    setVariable: (name, value) => {
      vars[name] = value;
    },

    getAll: () => vars,
  };
}

function run(src, vars = {}) {
  const scopes = makeScopes(vars);
  const vm = new VM(db, scopes);

  return new Promise(resolve => {
    vm.runSource(src, () => {
      expect(scopes.getAll()).toMatchSnapshot();
      resolve(undefined);
    });
  });
}

test('vm basic', async () => {
  return run(`=-(1 + 2 + 3)`, {
    number: x => {
      return x;
    },
    firstValue: arr => {
      return arr[0]['sum(t.amount)'];
    },
  });
});

test('vm boolean types', async () => {
  return run('=if(true and (1 + 2 + 3 - 5)) { 0 } else { 1 } ');
});
