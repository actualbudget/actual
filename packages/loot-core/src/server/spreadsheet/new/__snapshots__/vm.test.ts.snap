// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vm basic 1`] = `
Object {
  "firstValue": [Function],
  "generated!result": -6,
  "number": [Function],
}
`;

exports[`vm boolean types 1`] = `
Object {
  "generated!result": 0,
}
`;
